dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+              'end']:
+        return []
+
+    # Make sure we only request the period that was not retrieved before, for
+    # optimization
+    if start_timestamp < self.period_covered_by_retrieved_logs['start']:
+      if end_timestamp < self.period_covered_by_retrieved_logs['end']:
+        end_timestamp = self.period_covered_by_retrieved_logs['start']
+      self.period_covered_by_retrieved_logs['start'] = start_timestamp
+
+    if end_timestamp > self.period_covered_by_retrieved_logs['end']:
+      if start_timestamp > self.period_covered_by_retrieved_logs['start']:
+        start_timestamp = self.period_covered_by_retrieved_logs['end']
+      self.period_covered_by_retrieved_logs['end'] = end_timestamp
+
+    # pylint: disable=line-too-long
+    self.PublishMessage(
+        f'Retrieving logs from { start_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")} to {end_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}.'
+    )
+
+    query_filter = f'resource.type = ("gce_instance" OR "api" OR "gce_disk" OR "gce_image" OR "gce_instance_template" OR "gce_snapshot") \
+    AND logName = "projects/{project_id}/logs/cloudaudit.googleapis.com%2Factivity" \
+    AND operation.first = "true" \
+    AND timestamp >= "{start_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}" \
+    AND timestamp <= "{end_timestamp.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")}" \
+    AND severity=NOTICE \
+    AND protoPayload.methodName : ("insert" OR "create" OR "delete")'
+
+    if resource_id:
+      query_filter = query_filter + \
+          f' AND (resource.labels.instance_id="{resource_id}" OR resource.labels.image_id="{resource_id}")'
+
+    log_messages: List[Dict[str,
+                            Any]] = gcl.ExecuteQuery(qfilter=[query_filter])
+
+    return log_messages
+
+  def _ParseLogMessagesFromFileContainer(
+      self, file_container: containers.File) -> None:
+    """Parses GCP Cloud log messages supplied in the file container and fill
+    the self._resource_dict with the result.
+
+    Args:
+      file_container: file container
+    """
+    if not file_container:
+      self.logger.error('File container is null')
+      return
+
+    if not file_container.path:
+      self.logger.error('File container path is null or empty')
+
+    with open(file_container.path, 'r') as input_file:
+      log_messages = []
+      file_content = input_file.readline()
+
+      if not file_content:
+        self.logger.error(f'The supplied file {file_container.path} is empty')
+        return
+
+      while file_content:
+        log_messages.append(json.loads(file_content))
+        file_content = input_file.readline()
+
+      self._ParseLogMessages(log_messages)
+
+  def _ParseLogMessages(self, log_messages: List[Dict[str, Any]]) -> None:
+    """Parses supplied GCP Cloud log messages and fills self._resource_dict with
+    the result.
```

Comment:
function description in one line. you can follow on that with a more detailed description in a separate paragraph (with one blank line in-between)